### ISSUE TITLE
Error "Maximum number of clients reached" when 256 VizDoom instances started simultaneously

### ISSUE THREAD

**Issue by alex-petrenko** [mwydmuch/ViZDoom#364]
For RL project that uses a PPO algorithm, I am trying to run multiple VizDoom instances at the same time, distributed across 16-32 processes. With other envs I used to run ~256-512 environments at once, which allows me to generate actions for all of them in a single batch, therefore experience collection is very fast. But with VizDoom it causes problems.

This is what I see in the console when I create my workers:

```
Maximum number of clients reachedxcb_connection_has_error() returned true
Maximum number of clients reachedxcb_connection_has_error() returned true
Maximum number of clients reachedMaximum number of clients reachedxcb_connection_has_error() returned true
Maximum number of clients reached                                                                                                          
(vizdoom:16453): dbind-WARNING **: 22:53:21.795: Could not open X display
Maximum number of clients reachedxcb_connection_has_error() returned true
Maximum number of clients reachedMaximum number of clients reachedMaximum number of clients reachedxcb_connection_has_error() returned true
Maximum number of clients reachedxcb_connection_has_error() returned true                                                                  
Maximum number of clients reachedMaximum number of clients reachedMaximum number of clients reachedMaximum number of clients reachedMaximum number of clients reachedMaximum number of clients reachedMaximum number of clients reachedMaximum number of clients reachedxcb_connect
ion_has_error() returned true                                                                                                                                                                                                                                                  
Maximum number of clients reachedMaximum number of clients reachedxcb_connection_has_error() returned true
Maximum number of clients reachedMaximum number of clients reachedMaximum number of clients reachedxcb_connection_has_error() returned true
Maximum number of clients reachedxcb_connection_has_error() returned true
Maximum number of clients reachedxcb_connection_has_error() returned true
```
With 256 environments it keeps spitting out these messages for ~30 seconds, but then it seems like eventually, it works. But with more environments, it just gets stuck printing these messages. It feels like there's some spinlock that all of the processes are trying to acquire and it never happens.
The problem does not reproduce if I run 128 environments.

Does anyone know how to diagnose/fix this? Did anyone manage to run more than 256 Dooms on a single server?

* ViZDoom version you are using: 1.1.6
* information about your OS platform, distribution and environment: VizDoom installed from pip, Ubuntu 18.04, Python 3.6.7
* Configuration of the ViZDoom environment: resolution 160x120, scenario VizdoomMyWayHome-v0

**Comment by alex-petrenko** [comment]
@mwydmuch this is so cool, I really appreciate it a lot! I was planning to dig into code myself to find the problem. I will be looking at how you fixed it, very curious.

Thanks again, I will test as soon as I can!

**Comment by alex-petrenko** [comment]
Sorry, I forgot to close.
For some time I could not verify if the problems are completely gone, because after upgrading I started to get memory leaks. But I was able to verify that it's not VizDoom, but rather numpy 1.16.0 has some bugs that cause memory leaks.
After I figured all of that and rebooted, I don't see any problems anymore! Was able to run up to 1024 Dooms in parallel, and it works fine! Thanks for great work.
